File: task2_3.cpp
#include <locale.h>
#include <iostream>
using namespace std;

// Функция для нахождения обратного элемента
int extendedEuclideanAlgorithm(int number, int module, int& x, int& y)
{
    if (module == 0) {
        x = 1;
        y = 0;
        return number;
    }
    int x1 = 0, y1 = 0;
    int NOD = extendedEuclideanAlgorithm(module, number % module, x1, y1); // в реккуретной форме переставляем число и модуль местами
    x = y1;
    y = x1 - (number / module) * y1;
    return NOD;
}

int findInverseElement(int number, int module)
{
    int x = 0, y = 0;
    int NOD = extendedEuclideanAlgorithm(number, module, x, y);
    if (NOD != 1) {
        cout << "Обратного элемента не существует";
        return 0;
    }
    else {
        return (x % module + module) % module; // обработка отрицательных значений
    }
}

int main() {
    setlocale(LC_ALL, "Rus");
    int number = 0;
    int module = 0;
    cout << "Введите число" << endl;
    cin >> number;
    cout << "Введите модуль" << endl;
    cin >> module;
    int inverse_number = findInverseElement(number, module);
    cout << number << "^-1" << " mod " << module << " = " << inverse_number << endl;
    cout << number << " * " << inverse_number << " по модулю " << module << " = 1" << endl;
    return 0;
}

File: Go/task2_3.go
package main

import "fmt"

// Функция для нахождения обратного элемента
func extendedEuclideanAlgorithm(number, module int) (int, int, int) {
	if module == 0 {
		return number, 1, 0
	}
	NOD, x1, y1 := extendedEuclideanAlgorithm(module, number%module)
	x := y1
	y := x1 - (number/module)*y1
	return NOD, x, y
}

func findInverseElement(number, module int) int {
	NOD, x, _ := extendedEuclideanAlgorithm(number, module)
	if NOD != 1 {
		fmt.Println("Обратного элемента не существует")
		return 0
	} else {
		return (x%module + module) % module // обработка отрицательных значений
	}
}

func main() {
	var number, module int
	fmt.Println("Введите число")
	fmt.Scanln(&number)
	fmt.Println("Введите модуль")
	fmt.Scanln(&module)
	inverseNumber := findInverseElement(number, module)
	fmt.Printf("%d^-1 mod %d = %d\n", number, module, inverseNumber)
	fmt.Printf("%d * %d по модулю %d = 1\n", number, inverseNumber, module)
}
